os/board/common/crashdump.c

```diff
@@ -68,39 +138,50 @@ static int ramdump_via_uart(void)
 	size_t size;
 	uint8_t *ptr;
 	char host_reg[1] = "";
-	char *target_str = HANDSHAKE_STRING;
+	char *target_str = RAMDUMP_HANDSHAKE_STRING;
 	char host_buf[HANDSHAKE_STR_LEN_MAX] = "";
+	uint32_t userfs_start;
 
 #if !defined(CONFIG_ARCH_LOWPUTC)
-	/* If lowlevel serial is not available, ramdump is not possible */
+	/* If lowlevel serial is not available, dumping is not possible */
 	return -1;
 #endif
 
 	/* Inform the terminal user */
 	up_puts("\n************************************************************************************************\n");
-	up_puts("\t\tDisconnect this serial terminal and Run Ramdump Tool\n");
+	up_puts("\t\tDisconnect this serial terminal and run Dump Tool\n");
 	up_puts("************************************************************************************************\n");
+	if (CONFIG_FLASH_START_ADDR == 0xFFFFFFFF) {
+		up_puts("\nFLASH START ADDRESS IS INCORRECT!!! Userfs Dump will not work!!!\n");
+	}
 
-	/* Receive hanshake string from HOST */
+	/* Receive handshake string from HOST */
 	do {
 		host_buf[0] = up_getc();
 	} while (host_buf[0] != target_str[0]);
 
-	for (i = 1; i < strlen(target_str);) {
+	for (i = 1; i < HANDSHAKE_STR_LEN_MAX; ) {
 		if ((ch = up_getc()) != -1) {
 			host_buf[i] = ch;
 			i++;
 		}
 	}
 
-	if (strncmp(host_buf, target_str, strlen(target_str)) != 0) {
+	/* Check handshake string against RAMDUMP HANDSHAKE string */
```

Comment:
crashdump.c
```
#define RAMDUMP_STR "TIZENRTRMDUMP"
#define FSDUMP_STR     "TIZENRTFSDUMP"
#define REBOOT_STR      "TIZENRTREBOOT"

up_puts("\n-------\n");
up_puts("\t\tDisconnect .....\n");
up_puts("\n-------\n");
while (1) {
    do {
        host_buf[0] = up_getc();
    } while (host_buf[0] != 'T');

    for (i = 1; i < 13; ) {
        ....
    }

    if (!strncmp(host_buf, TIZENRTRMDUMP, 14)) {
        do_ramdump();
    } else if (TIZENRFSDUMP) {
        do_fsdump();
    } else if (TIZENRTREBOOT) {
        do_reboot();
    } else {
        up_lowputc('N');
    }
}
```
dump_tool.c
```
printf("Target device locked and ready to DUMP!!!\n");

while (1) {
	printf("Chose from the following options:-\n1. RAM Dump\n2. Userfs Dump\n3. Both Dumps\n4. Exit Dump Tool\n");
	fflush(stdout);
	scanf("%d", &choice);
	getchar();

        switch (choice) {
        case 1:
             do_ramdump = true;
              break;
        case 2:
              do_fsdump = true;
              break;
         case 3:
              do_reboot = true;
              break;
         case 4:
               do_ramdump = do_fsdump = true;
               break;
         }

         if (do_ramdump) {
               get_ramdump();
         }

         if (do_fsdump) {
                 get_fsdump();
         }

          if (exit) {
               printf(".....");
               return;
          }
          .....
}
```

I suggest the device stays in while loop before reboot command to allow getting dump after some time.

Reply:
In case of TIZENRTREBOOT, do_reboot() is to be called according to your suggestion.
Here, the function simply causes the device to return and exit crashdump.c right?

Reply:
> if (exit) {
>                printf(".....");
>                return;
>           }
>           .....

by "exit" you mean "do_reboot"?

Reply:
The reboot means device power reset.
```
boardctl(BOARDIOC_RESET, EXIT_SUCCESS);
```
This is the architecture-specific code which is working with watchdog.

The exit means just exiting the tool.
When I think the usages of crash, QA team keeps crash state and gets dump with request.
So, QA team executes the tool and gets dumps and terminates the tool. But the device is in the loop to get additional dump request for later.

Reply:
So, the device gets 3 commands, ramdump, fsdump and reboot. The tool gets 4 commands, ramdump, fsdump, reboot and exit.

Reply:
Updated